File: dotnet-backend/src/eTour.Application/Payment/CardDetails.cs
using System.Text.RegularExpressions;

namespace eTour.Application.Gateways;

/// <summary>Card data passed to a gateway for a single charge. Deliberately not persisted anywhere.</summary>
public partial class CardDetails
{
    private const string DeclineCardPrefix = "4000000000000002";

    public CardDetails(string? number, string holderName, int expiryMonth, int expiryYear, string cvv)
    {
        Number = number is null ? null : WhitespaceOrDash().Replace(number, "");
        HolderName = holderName;
        ExpiryMonth = expiryMonth;
        ExpiryYear = expiryYear;
        Cvv = cvv;
    }

    public string? Number { get; }
    public string HolderName { get; }
    public int ExpiryMonth { get; }
    public int ExpiryYear { get; }
    public string Cvv { get; }

    /// <summary>Last four digits - the only part of the number safe to keep.</summary>
    public string Last4 => Number is null || Number.Length < 4 ? "____" : Number[^4..];

    /// <summary>Brand inferred from the IIN/BIN prefix. Presentational only.</summary>
    public string Brand
    {
        get
        {
            if (string.IsNullOrEmpty(Number))
            {
                return "CARD";
            }
            if (Number.StartsWith('4'))
            {
                return "VISA";
            }
            if (MastercardPrefix().IsMatch(Number))
            {
                return "MASTERCARD";
            }
            if (AmexPrefix().IsMatch(Number))
            {
                return "AMEX";
            }
            if (Number.StartsWith('6'))
            {
                return "DISCOVER";
            }
            if (Number.StartsWith("35"))
            {
                return "JCB";
            }
            if (RupayPrefix().IsMatch(Number))
            {
                return "RUPAY";
            }
            return "CARD";
        }
    }

    /// <summary>e.g. "VISA ****4242" - safe to store and display.</summary>
    public string MaskedSummary => $"{Brand} ****{Last4}";

    /// <summary>Luhn check digit validation.</summary>
    public bool PassesLuhn()
    {
        if (Number is null || !DigitsOnly12To19().IsMatch(Number))
        {
            return false;
        }

        var sum = 0;
        var doubleIt = false;
        for (var i = Number.Length - 1; i >= 0; i--)
        {
            var digit = Number[i] - '0';
            if (doubleIt)
            {
                digit *= 2;
                if (digit > 9)
                {
                    digit -= 9;
                }
            }
            sum += digit;
            doubleIt = !doubleIt;
        }
        return sum % 10 == 0;
    }

    /// <summary>A card is valid through the LAST day of its expiry month (YearMonth comparison, not exact day).</summary>
    public bool IsExpired(DateOnly asOf)
    {
        if (ExpiryMonth is < 1 or > 12)
        {
            return true;
        }
        return ExpiryYear * 12 + ExpiryMonth < asOf.Year * 12 + asOf.Month;
    }

    /// <summary>AMEX uses a 4-digit CID; everything else uses 3.</summary>
    public bool HasValidCvv()
    {
        if (Cvv is null)
        {
            return false;
        }
        var expected = Brand == "AMEX" ? 4 : 3;
        return Cvv.Length == expected && Cvv.All(char.IsDigit);
    }

    public override string ToString() => $"CardDetails{{{MaskedSummary}}}";

    [GeneratedRegex(@"\s|-")]
    private static partial Regex WhitespaceOrDash();

    [GeneratedRegex("^5[1-5]|^2[2-7]")]
    private static partial Regex MastercardPrefix();

    [GeneratedRegex("^3[47]")]
    private static partial Regex AmexPrefix();

    [GeneratedRegex("^(60|65|81|82|508)")]
    private static partial Regex RupayPrefix();

    [GeneratedRegex(@"^\d{12,19}$")]
    private static partial Regex DigitsOnly12To19();
}

File: Backend/src/main/java/com/etour/payment/CardDetails.java
package com.etour.payment;

import java.time.YearMonth;

/**
 * Card data passed to a gateway for a single charge. Deliberately NOT an
 * entity and never persisted.
 *
 * {@link #toString()} is overridden so a stray log statement or an exception
 * stack trace can never print a PAN or CVV.
 */
public class CardDetails {

    private final String number;
    private final String holderName;
    private final int expiryMonth;
    private final int expiryYear;
    private final String cvv;

    public CardDetails(String number, String holderName, int expiryMonth, int expiryYear, String cvv) {
        this.number = number == null ? null : number.replaceAll("\\s|-", "");
        this.holderName = holderName;
        this.expiryMonth = expiryMonth;
        this.expiryYear = expiryYear;
        this.cvv = cvv;
    }

    public String getNumber() { return number; }
    public String getHolderName() { return holderName; }
    public int getExpiryMonth() { return expiryMonth; }
    public int getExpiryYear() { return expiryYear; }
    public String getCvv() { return cvv; }

    /** Last four digits - the only part of the number safe to keep. */
    public String getLast4() {
        if (number == null || number.length() < 4) return "____";
        return number.substring(number.length() - 4);
    }

    /**
     * Brand inferred from the IIN/BIN prefix. Presentational only - the
     * gateway is the authority on whether a card is actually usable.
     */
    public String getBrand() {
        if (number == null || number.isEmpty()) return "CARD";
        if (number.startsWith("4")) return "VISA";
        if (number.matches("^5[1-5].*") || number.matches("^2[2-7].*")) return "MASTERCARD";
        if (number.matches("^3[47].*")) return "AMEX";
        if (number.startsWith("6")) return "DISCOVER";
        if (number.startsWith("35")) return "JCB";
        if (number.matches("^(60|65|81|82|508).*")) return "RUPAY";
        return "CARD";
    }

    /** e.g. "VISA ****4242" - safe to store and display. */
    public String getMaskedSummary() {
        return getBrand() + " ****" + getLast4();
    }

    /** Luhn check digit validation. */
    public boolean passesLuhn() {
        if (number == null || !number.matches("\\d{12,19}")) return false;
        int sum = 0;
        boolean doubleIt = false;
        for (int i = number.length() - 1; i >= 0; i--) {
            int digit = number.charAt(i) - '0';
            if (doubleIt) {
                digit *= 2;
                if (digit > 9) digit -= 9;
            }
            sum += digit;
            doubleIt = !doubleIt;
        }
        return sum % 10 == 0;
    }

    /** A card is valid through the LAST day of its expiry month. */
    public boolean isExpired(YearMonth now) {
        if (expiryMonth < 1 || expiryMonth > 12) return true;
        return YearMonth.of(expiryYear, expiryMonth).isBefore(now);
    }

    /** AMEX uses a 4-digit CID; everything else uses 3. */
    public boolean hasValidCvv() {
        if (cvv == null) return false;
        int expected = "AMEX".equals(getBrand()) ? 4 : 3;
        return cvv.matches("\\d{" + expected + "}");
    }

    @Override
    public String toString() {
        // Never expose the PAN or CVV.
        return "CardDetails{" + getMaskedSummary() + "}";
    }
}
